### Q1. Pandas version

In [44]:
import pandas as pd
import numpy as np

In [4]:
pd.__version__

'2.3.1'

#### Getting the data

For this homework, we will download the Car Fuel Efficiency dataset:

In [7]:
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv

--2025-09-27 14:44:51--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 874188 (854K) [text/plain]
Saving to: ‘car_fuel_efficiency.csv’

car_fuel_efficiency 100%[===================>] 853.70K  --.-KB/s    in 0.01s   

2025-09-27 14:44:51 (73.6 MB/s) - ‘car_fuel_efficiency.csv’ saved [874188/874188]



In [8]:
car_fuel = pd.read_csv('car_fuel_efficiency.csv')

First 5 rows of the dataset:

In [9]:
car_fuel.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

### Q2. Records count

In [13]:
car_fuel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


How many records are in the dataset? **9704** 

### Q3. Fuel types

In [15]:
car_fuel['fuel_type'].unique()

array(['Gasoline', 'Diesel'], dtype=object)

How many fuel types are presented in the dataset? **2**

### Q4. Missing values

Count missing values per column:

In [16]:
car_fuel.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

In [ ]:
Filter only columns with missing values

In [21]:
num_cols_with_missing = (car_fuel.isnull().sum() > 0).sum()
print(f"How many columns in the dataset have missing values? {num_cols_with_missing}")

How many columns in the dataset have missing values? 4


### Q5. Max fuel efficiency

In [27]:
round(float(car_fuel[car_fuel['origin']=='Asia']['fuel_efficiency_mpg'].max()),2)

23.76

What's the maximum fuel efficiency of cars from Asia? **23.76**

### Q6. Median value of horsepower

1. Find the median value of `horsepower` column in the dataset.

In [29]:
median_horsepower = car_fuel['horsepower'].median()
float(median_horsepower)

149.0

2. Next, calculate the most frequent value of the same `horsepower` column.

In [32]:
mode_horsepower = car_fuel['horsepower'].mode()[0]
float(mode_horsepower)

152.0

3. Use `fillna` method to fill the missing values in `horsepower` column with the most frequent value from the previous step.

In [33]:
car_fuel['horsepower'] = car_fuel['horsepower'].fillna(mode_horsepower)

4. Now, calculate the median value of `horsepower` once again.

In [34]:
updated_median_horsepower = car_fuel['horsepower'].median()
float(updated_median_horsepower)

152.0

Has it changed? **Yes, it increased**

### Q7. Sum of weights

1. Select all the cars from Asia

In [38]:
df = car_fuel[car_fuel['origin']=='Asia']
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                   250            1.0       174.0     2714.219310   
12                  320            5.0       145.0     2783.868974   
14                  200            6.0       160.0     3582.687368   
20                  150            3.0       197.0     2231.808142   
21                  160            4.0       133.0     2659.431451   

    acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8           10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12          15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14          14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20          18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21           NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   

    fuel_efficiency_mpg  
8             16.823554  
12            16.175820  
14            11.871091  
20            18.889083  
21            16.077730

2. Select only columns `vehicle_weight` and `model_year`

In [39]:
df = df[['vehicle_weight', 'model_year']]
df.head()

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016

3. Select the first 7 values

In [41]:
df = df.head(7)
df

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

4. Get the underlying NumPy array. Let's call it `X`.

In [42]:
X = df.to_numpy()
X

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ],
       [2844.22753389, 2014.        ],
       [3761.99403819, 2019.        ]])

5. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.

In [45]:
XTX = X.T @ X
XTX

array([[62248334.33150762, 41431216.5073268 ],
       [41431216.5073268 , 28373339.        ]])

6. Invert `XTX`

In [46]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 5.71497081e-07, -8.34509443e-07],
       [-8.34509443e-07,  1.25380877e-06]])

7. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100, 1200]`.

In [48]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

8. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.

In [51]:
w = XTX_inv @ X.T @ y
w

array([0.01386421, 0.5049067 ])

9. What's the sum of all the elements of the result?

9. What's the sum of all the elements of the result?

In [57]:
result = w.sum()
round(float(result),2)

0.52